core/modules/user.py

```diff
@@ -751,28 +750,41 @@ def secondFactorSucceeded(self, secondFactor, userKey):
             raise errors.RequestTimeout()
         return self.authenticateUser(userKey)
 
-    def authenticateUser(self, userKey: db.Key, **kwargs):
+    def authenticateUser(self, key: db.Key, **kwargs):
         """
-            Performs Log-In for the current session and the given userKey.
+            Performs Log-In for the current session and the given user key.
 
             This resets the current session: All fields not explicitly marked as persistent
             by conf["viur.session.persistentFieldsOnLogin"] are gone afterwards.
 
-            :param userKey: The (DB-)Key of the user we shall authenticate
+            :param key: The (DB-)Key of the user we shall authenticate
         """
-        currSess = utils.currentSession.get()
-        res = db.Get(userKey)
-        assert res, "Unable to authenticate unknown user %s" % userKey
-        oldSession = {k: v for k, v in currSess.items()}  # Store all items in the current session
-        currSess.reset()
-        # Copy the persistent fields over
-        for k in conf["viur.session.persistentFieldsOnLogin"]:
-            if k in oldSession:
-                currSess[k] = oldSession[k]
-        del oldSession
-        currSess["user"] = res
-        currSess.markChanged()
-        utils.currentRequest.get().response.headers["Sec-X-ViUR-StaticSKey"] = currSess.staticSecurityKey
+        skel = self.baseSkel()
+        if not skel.fromDB(key):
+            raise ValueError(f"Unable to authenticate unknown user {key}")
+
+        # Update the lastlogin timestamp (if available!)
+        if "lastlogin" in skel:
+            now = utils.utcNow()
+
+            # Conserve DB-Writes: Update the user max once in 30 Minutes (why??)
+            if not skel["lastlogin"] or ((now - skel["lastlogin"]) > datetime.timedelta(minutes=30)):
+                skel["lastlogin"] = now
+                skel.toDB()
+
+        # Update session for user
+        session = utils.currentSession.get()
+        # Remember persistent fields...
+        take_over = {k: v for k, v in session.items() if k in conf["viur.session.persistentFieldsOnLogin"]}
+        session.reset()
+        # and copy them over to the new session
+        session |= take_over
+
+        session["user"] = skel.dbEntity
+        session.markChanged()
+
+        utils.currentRequest.get().response.headers["Sec-X-ViUR-StaticSKey"] = session.staticSecurityKey
+
         self.onLogin()
```

Comment:
I stumbled over this already during my PR. Shouldn't we pass the user skel directly to the `onLogin` method to save the `getCurrentUser` and the related Skel-creation/`db.Get` there?

```suggestion
        self.onLogin(skel)
```
(and extend the signature of onLogin).

Reply:
Indeed, this makes sense. Otherwise, you might break existing projects when they don't ask for `*args`, which should be normal when sub-classing the User module.

Anyway, I think currently every minor release should also be carefully introduced, and when it's broken, it will directly break when logging in. It would also make the signatures of `onLogin` and `onLogout` equal.